lib/mss/content_protection.js

```diff
@@ -0,0 +1,338 @@
+        protectionHeader, keySystemsBySystemId);
+  }
+
+  /**
+   * Parses an Array buffer starting at byteOffset for PlayReady Object Records.
+   * Each PRO Record is preceded by its PlayReady Record type and length in
+   * bytes.
+   *
+   * PlayReady Object Record format: https://goo.gl/FTcu46
+   *
+   * @param {!DataView} view
+   * @param {number} byteOffset
+   * @return {!Array.<shaka.mss.ContentProtection.PlayReadyRecord>}
+   * @private
+   */
+  static parseMsProRecords_(view, byteOffset) {
+    const records = [];
+
+    while (byteOffset < view.byteLength - 1) {
+      const type = view.getUint16(byteOffset, true);
+      byteOffset += 2;
+
+      const byteLength = view.getUint16(byteOffset, true);
+      byteOffset += 2;
+
+      if ((byteLength & 1) != 0 || byteLength + byteOffset > view.byteLength) {
+        shaka.log.warning('Malformed MS PRO object');
+        return [];
+      }
+
+      const recordValue = shaka.util.BufferUtils.toUint8(
+          view, byteOffset, byteLength);
+      records.push({
+        type: type,
+        value: recordValue,
+      });
+
+      byteOffset += byteLength;
+    }
+
+    return records;
+  }
+
+  /**
+   * Parses a buffer for PlayReady Objects.  The data
+   * should contain a 32-bit integer indicating the length of
+   * the PRO in bytes.  Following that, a 16-bit integer for
+   * the number of PlayReady Object Records in the PRO.  Lastly,
+   * a byte array of the PRO Records themselves.
+   *
+   * PlayReady Object format: https://goo.gl/W8yAN4
+   *
+   * @param {BufferSource} data
+   * @return {!Array.<shaka.mss.ContentProtection.PlayReadyRecord>}
+   * @private
+   */
+  static parseMsPro_(data) {
+    let byteOffset = 0;
+    const view = shaka.util.BufferUtils.toDataView(data);
+
+    // First 4 bytes is the PRO length (DWORD)
+    const byteLength = view.getUint32(byteOffset, /* littleEndian= */ true);
+    byteOffset += 4;
+
+    if (byteLength != data.byteLength) {
+      // Malformed PRO
+      shaka.log.warning('PlayReady Object with invalid length encountered.');
+      return [];
+    }
+
+    // Skip PRO Record count (WORD)
+    byteOffset += 2;
+
+    // Rest of the data contains the PRO Records
+    const ContentProtection = shaka.mss.ContentProtection;
+    return ContentProtection.parseMsProRecords_(view, byteOffset);
+  }
+
+  /**
+   * PlayReady Header format: https://goo.gl/dBzxNA
+   *
+   * @param {!Element} xml
+   * @return {string}
+   * @private
+   */
+  static getLaurl_(xml) {
+    // LA_URL element is optional and no more than one is
+    // allowed inside the DATA element. Only absolute URLs are allowed.
+    // If the LA_URL element exists, it must not be empty.
+    for (const elem of xml.getElementsByTagName('DATA')) {
+      for (const child of elem.childNodes) {
+        if (child instanceof Element && child.tagName == 'LA_URL') {
+          return child.textContent;
+        }
+      }
+    }
+
+    // Not found
+    return '';
```

Comment:
Is it possible for a `LA_URL` element to be empty, or would that be considered invalid content? If the `LA_URL` could be empty, the fallback return should probably be null instead.

Reply:
I added a comment